File: src/emitter/mermaid/push-edge-lines.test.ts
import { describe, expect, test } from "vitest";

import { pushEdgeLines } from "./push-edge-lines.js";
import { makeEdge } from "./testing/make-edge.js";

describe("pushEdgeLines", () => {
  test("formats each edge as '  <from> -->|<label>| <to>'", () => {
    const lines: /* mutable */ string[] = [];
    pushEdgeLines(
      [
        makeEdge({ from: "a", to: "b", label: "read" }),
        makeEdge({ from: "c", to: "d", label: "write" }),
      ],
      lines,
    );
    expect(lines).toEqual(["  a -->|read| b", "  c -->|write| d"]);
  });

  test("preserves the input order in the appended lines", () => {
    const lines: /* mutable */ string[] = ["existing"];
    pushEdgeLines([makeEdge({ from: "x", to: "y", label: "set" })], lines);
    expect(lines).toEqual(["existing", "  x -->|set| y"]);
  });

  test("empty input pushes nothing", () => {
    const lines: /* mutable */ string[] = [];
    pushEdgeLines([], lines);
    expect(lines).toEqual([]);
  });
});

File: crates/unsnarl-emitter-mermaid/src/push_edge_lines_test.rs
use unsnarl_visual_graph::visual_edge::VisualEdge;

use super::push_edge_lines;

fn edge(from: &str, to: &str, label: &str) -> VisualEdge {
    VisualEdge::new(from, to, label)
}

#[test]
fn formats_each_edge_as_arrow_with_label() {
    let edges = vec![edge("a", "b", "read"), edge("c", "d", "write")];
    let mut lines: Vec<String> = Vec::new();
    push_edge_lines(&edges, &mut lines, None);
    assert_eq!(
        lines,
        vec![
            "  a -->|read| b".to_string(),
            "  c -->|write| d".to_string(),
        ]
    );
}

#[test]
fn preserves_input_order_in_the_appended_lines() {
    let edges = vec![edge("x", "y", "set")];
    let mut lines: Vec<String> = vec!["existing".to_string()];
    push_edge_lines(&edges, &mut lines, None);
    assert_eq!(
        lines,
        vec!["existing".to_string(), "  x -->|set| y".to_string()]
    );
}

#[test]
fn empty_input_pushes_nothing() {
    let edges: Vec<VisualEdge> = Vec::new();
    let mut lines: Vec<String> = Vec::new();
    push_edge_lines(&edges, &mut lines, None);
    assert!(lines.is_empty());
}
